### PR TITLE
cmake: zephyr_file() now accepts BOARD and BOARD_REVISION as argument

### PR DESCRIPTION
**Pull request by tejlmand**
With the introduction of #32556 changing the BOARD now prints a warning.

Zephyr provides `zephyr_file()` to look up overlays and Kconfig
fragments for a specified board, and it used to be possible and safe to
do:
```
function(my_func)
  set(BOARD <local_scope_board>)
  zephyr_file(CONF_FILES ...)
endfunction(my_func)
```

As the BOARD inside `my_func` is locally scoped, this is safe to do.
But with introduction of #32556 a warning is now printed when running
CMake.

Therefore `zephyr_file` has been extended to allow for optional BOARD
and BOARD_REVISION arguments.
If BOARD is not given as argument the current BOARD in the build system
will be used.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

### CODE REVIEW

File: cmake/extensions.cmake [hunk 1]
```diff
@@ -1937,10 +1944,26 @@ Relative paths are only allowed with `-D${ARGV1}=<path>`")
   endif()
 
   if(FILE_CONF_FILES)
-    set(FILENAMES ${BOARD})
+    if(DEFINED FILE_BOARD_REVISION AND NOT FILE_BOARD)
+        message(FATAL_ERROR
+	  "zephyr_file(${ARGV0} <path> BOARD_REVISION ${FILE_BOARD_REVISION} ...)"
+	  " given without BOARD argument, please specify BOARD"
+	)
```
Comment (anangl) [inline cmake/extensions.cmake:1951]: This indentation is misleading.

Comment (tejlmand) [reply cmake/extensions.cmake]: Thanks for spotting.
A couple of stray tabs that has now been converted to spaces.

**Review by anangl — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
> Zephyr provides `zephyr_file()` to look up overlays and Kconfig
> fragments for a specified board, and it used to be possible and safe to
> do:

I'm tempted to say "what valid upstream use cases are there for doing such a crazy thing" :P

But sure, this seems like a valid workaround for our downstream issue.

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by tejlmand** [comment]
> I'm tempted to say "what valid upstream use cases are there for doing such a crazy thing" :P
> 
> But sure, this seems like a valid workaround for our downstream issue.

As of now, none. But TF-M could be a candidate where board would be changed in a local scope.

But with the introducing of #32556 we are preventing everyone from setting BOARD in a locally scoped environment, which were  not the intention of #32556.
The intention was to help developers to not set board by a mistake, when it is not intended.

Unfortunately this prevents users from setting a locally scoped BOARD variable for other reasons.

So the only two options I can think of, is:
- Reverting #32556, so that users can still set BOARD as a locally scoped variable.
- This PR.

but as I like the warning when people sets `BOARD` I prefer to not go back to the state before.
But then we need to provide something else to users.

**Pull request merged**